firefox_puppeteer/api/places.py

```diff
@@ -0,0 +1,156 @@
+
+          let bookmarkIds = bs.getBookmarkIdsForURI(ios.newURI(url, null, null), {});
+          let folderIds = [];
+
+          for (let i = 0; i < bookmarkIds.length; i++) {
+            folderIds.push(bs.getFolderIdForItem(bookmarkIds[i]));
+          }
+
+          return folderIds;
+        """, script_args=[url])
+
+    def is_bookmark_star_button_ready(self):
+        """Checks if the status of the star-button is not updating.
+
+        :returns: True, if the button is ready
+        """
+        return self.marionette.execute_script("""
+          let button = window.BookmarkingUI;
+
+          return button.status !== button.STATUS_UPDATING;
+        """)
+
+    def restore_default_bookmarks(self):
+        """Restores the default bookmarks for the current profile."""
+        try:
+            return self.marionette.execute_async_script("""
+              Cu.import("resource://gre/modules/BookmarkHTMLUtils.jsm");
+              Cu.import("resource://gre/modules/Services.jsm");
+
+              // Default bookmarks.html file is stored inside omni.jar,
+              // so get it via a resource URI
+              let defaultBookmarks = 'resource:///defaults/profile/bookmarks.html';
+
+              let observer = {
+                observe: function (aSubject, aTopic, aData) {
+                  Services.obs.removeObserver(observer, "bookmarks-restore-success");
+
+                  marionetteScriptFinished(true);
+                }
+              };
+
+              // Fire off an initially import
+              Services.obs.addObserver(observer, "bookmarks-restore-success", false);
+              BookmarkHTMLUtils.importFromURL(defaultBookmarks, true);
+            """, script_timeout=10000)
+        except TimeoutException:
+            # TODO: In case of a timeout clean-up the registered topic
+            pass
+
+    # Browser history related helpers #
+
+    def remove_all_history(self):
+        """Removes all history items."""
+        try:
+            self.marionette.execute_async_script("""
+                Cu.import("resource://gre/modules/Services.jsm");
+
+                let hs = Cc["@mozilla.org/browser/nav-history-service;1"]
+                         .getService(Ci.nsIBrowserHistory);
+
+                let observer = {
+                  observe: function (aSubject, aTopic, aData) {
+                    Services.obs.removeObserver(observer, 'places-expiration-finished');
+
+                    marionetteScriptFinished(true);
+                  }
+                };
+
+                // Remove the pages, then block until we're done or until timeout is reached
+                Services.obs.addObserver(observer, 'places-expiration-finished', false);
+
+                hs.removeAllPages();
+            """, script_timeout=10000)
+        except TimeoutException:
+            # TODO: In case of a timeout clean-up the registered topic
+            pass
+
+    def wait_for_visited(self, urls, callback):
```

Comment:
If it's just a matter of keeping the observer around to unregister it, isn't this an instance we can use execute script with newSandbox=False and stash the observer? I wouldn't push this, but I really think it's preferable to adding a 4 second sleep, which will make developing and running these tests significantly less convenient.